src/vm/eventpipe.cpp

```diff
@@ -460,19 +490,26 @@ void EventPipe::DisableInternal(EventPipeSessionID id, EventPipeProviderCallback
 EventPipeSession *EventPipe::GetSession(EventPipeSessionID id)
 {
     LIMITED_METHOD_CONTRACT;
-    _ASSERTE(IsLockOwnedByCurrentThread());
+    _ASSERTE(s_tracingInitialized);
 
-    if (s_pSessions == nullptr)
+    if (!s_tracingInitialized)
         return nullptr;
 
-    EventPipeSession *pSession = nullptr;
-    return s_pSessions->Lookup(id, &pSession) ? pSession : nullptr;
+    // Attempt to get the specified session ID.
+    for (uint64_t i = 0; i < MaxNumberOfSessions; ++i)
+    {
+        EventPipeSession *const pSession = s_pSessions[i].Load();
+        if (pSession != nullptr || pSession->GetId() == id)
+            return pSession;
+    }
+
+    return nullptr;
 }
 
 bool EventPipe::Enabled()
 {
     LIMITED_METHOD_CONTRACT;
-    return s_tracingInitialized && (s_pSessions != NULL) && (s_pSessions->GetCount() > 0);
+    return s_tracingInitialized && (s_pConfig != nullptr) && s_pConfig->Enabled();
```

Comment:
s_pConfig gets deleted in EventPipe::Shutdown(), which could occur right after the check for s_pConfig != nullptr and right before s_pConfig->Enabled(). If that happens this will AV. You could resolve this by making the enabled flag a global or making EventPipeConfiguration a global value rather than a global ptr (ie s_config instead of s_pConfig)

Reply:
I like the idea of having a global value instead of a dynamic pointer :)